### PR TITLE
feat: make read_group and read_window_aggregate work across chunks

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/812

Note: I have taken a slightly different approach with this PR than in previous porting exercises -- I have simply brought the code from the mutable buffer over without removing the old version yet. I plan to remove the old implementations with several follow on PRs and thus spare you, dear reviewer, an even larger PR

# Rationale:
Users want to query data via flux / influxql when it lives in any type of IOx chunk. Flux / InfuxQL use the gRPC storage API.

# Changes
This PR ports the `read_group` and `read_window_aggregate`(which is very similar plan) into the `InfluxRPCPlanner` from the `MutableBuffer` so that it works across all types of Chunks.

# Follow on work:

## Remove old version in the MutableBuffer:

This removal of the vesigal query code from the mutable buffer is tracked in https://github.com/influxdata/influxdb_iox/issues/815. The work is roughly:

1. Move chunk predicate creation from `query::Predicate` into `server` crate and out of `mutable_buffer` crate
2. Move anything else needed out
3. Remove query dependency in mutable buffer and delete all the old code.
3. Remove  GroupByAndAggregate and call the right thing directly -- the combining read_group and read_window_aggregate to a single request just to dispatch again in InfluxRPC is silly, but I kept the structure in this PR to minimize the size of an already large PR.

## Better query testing harness
* https://github.com/influxdata/influxdb_iox/issues/843: for a data driven testing harness  (this is tantalizing close now that @tustvold has the management API almost ready to go)

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: query/src/frontend/influxrpc.rs [hunk 1]
```diff
@@ -521,23 +540,118 @@ impl InfluxRPCPlanner {
...
+                ss_plans.push(ss_plan);
+            }
+        }
+
+        Ok(ss_plans.into())
+    }
+
+    /// Query groups (dispatch to read_group or read_window_aggregate).
+    /// TODO remove this dispatch and just call the right thing directly
+    pub async fn query_group<D>(
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:565]: I can't really remember why I thought `GroupByAndAggregate`  would be a useful thing. I plan to remove it in a follow on PR and call the right functions directly

File: query/src/frontend/influxrpc.rs [hunk 2]
```diff
@@ -521,23 +540,118 @@ impl InfluxRPCPlanner {
...
+                    .await
+            }
+            GroupByAndAggregate::Window { agg, every, offset } => {
+                self.read_window_aggregate(database, predicate, agg, every, offset)
+                    .await
+            }
+        }
+    }
+
+    /// Creates a GroupedSeriesSet plan that produces an output table
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:586]: Here is the logic for `read_group`

File: query/src/frontend/influxrpc.rs [hunk 3]
```diff
@@ -521,23 +540,118 @@ impl InfluxRPCPlanner {
...
+                ss_plans.push(grouped_plan);
             }
         }
 
+        Ok(ss_plans.into())
+    }
+
+    /// Creates a GroupedSeriesSet plan that produces an output table with rows
+    /// that are grouped by window defintions
+    pub async fn read_window_aggregate<D>(
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:633]: Here is the logic for `read_window_aggregate `

File: query/src/frontend/influxrpc.rs [hunk 4]
```diff
@@ -548,6 +662,28 @@ impl InfluxRPCPlanner {
         Ok(ss_plans.into())
     }
 
+    /// Creates a map of table_name --> Chunks that have that table
+    async fn group_chunks_by_table<C>(
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:666]: this was just factored out so I could reuse the logic

File: query/src/frontend/influxrpc.rs [hunk 5]
```diff
@@ -747,7 +898,7 @@ impl InfluxRPCPlanner {
         // Select away anything that isn't in the influx data model
         let tags_fields_and_timestamps: Vec<Expr> = schema
             .tags_iter()
-            .chain(schema.fields_iter())
+            .chain(filtered_fields_iter(&schema, predicate))
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:901]: `filtered_fields_iter` is worth pointing out -- some of the end to end tests showed that predicates of the form `_field = "foo"` were being ignored in read_aggregate and upon further inspection, this turned out to affect read_filter as well

File: query/src/frontend/influxrpc.rs [hunk 6]
```diff
@@ -973,3 +1326,271 @@ struct TableScanAndFilter {
     /// The IOx schema of the result
     schema: Schema,
 }
+
+/// Reorders tag_columns so that its prefix matches exactly
+/// prefix_columns. Returns an error if there are duplicates, or other
+/// untoward inputs
+fn reorder_prefix<'a>(
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:1333]: This logic / tests is mostly moved from the mutable buffer (though I changed the type signature)

File: query/src/frontend/influxrpc.rs [hunk 7]
```diff
@@ -973,3 +1326,271 @@ struct TableScanAndFilter {
...
+    predicate: &'a Predicate,
+) -> impl Iterator<Item = &'a Field> {
+    schema
+        .fields_iter()
+        .filter(move |f| predicate.should_include_field(f.name()))
+}
+
+/// Creates aggregate expressions and tracks field output according to
+/// the rules explained on `read_group_plan`
+impl AggExprs {
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:1407]: likewise this code is all basically the same as was (actually still is) in the mutable_buffer, but it was hoisted into this crate and I plan to remove the mutable-buffer version shortly

File: query/src/lib.rs [hunk 8]
```diff
@@ -50,24 +49,6 @@ pub trait Database: Debug + Send + Sync {
     /// covering set means that together the chunks make up a single
     /// complete copy of the data being queried.
     fn chunks(&self, partition_key: &str) -> Vec<Arc<Self::Chunk>>;
-
-    // ----------
```
Comment (alamb) [inline query/src/lib.rs:54]: 🎉  here is the removal of `query_groups` from the chunk interface

File: server/src/query_tests/influxrpc/read_filter.rs [hunk 9]
```diff
@@ -57,96 +48,22 @@ macro_rules! run_read_filter_test_case {
...
-                scenario_name,
-                expected_results,
-                string_results
+                scenario_name, expected_results, string_results
             );
         }
     };
 }
 
-/// Format the field indexes into strings
```
Comment (alamb) [inline server/src/query_tests/influxrpc/read_filter.rs:112]: this stuff was pulled out into another module

File: server/src/query_tests/influxrpc/read_filter.rs [hunk 10]
```diff
@@ -246,6 +163,50 @@ async fn test_read_filter_data_filter() {
     run_read_filter_test_case!(TwoMeasurementsMultiSeries {}, predicate, expected_results);
 }
 
+#[tokio::test]
+async fn test_read_filter_data_filter_fields() {
+    // filter out one row in h20
+    let predicate = PredicateBuilder::default()
+        .field_columns(vec!["other_temp"])
```
Comment (alamb) [inline server/src/query_tests/influxrpc/read_filter.rs:170]: this is a new test for `read_filter` showing that the field filtering works

File: server/src/query_tests/influxrpc/read_group.rs [hunk 11]
```diff
@@ -0,0 +1,546 @@
+//! Tests for the Influx gRPC queries
```
Comment (alamb) [inline server/src/query_tests/influxrpc/read_group.rs:1]: These tests are ported from `mutable_buffer/src/table.rs`

File: server/src/query_tests/influxrpc/read_window_aggregate.rs [hunk 12]
```diff
@@ -0,0 +1,229 @@
+//! Tests for the Influx gRPC queries
```
Comment (alamb) [inline server/src/query_tests/influxrpc/read_window_aggregate.rs:1]: `mutable_buffer/src/table.rs`

File: src/influxdb_ioxd/rpc/storage/service.rs [hunk 13]
```diff
@@ -1643,13 +1639,8 @@ mod tests {
...
             .to_string();
-        let expected_error = "Sugar we are going down";
-        assert!(
-            response_string.contains(expected_error),
-            "'{}' did not contain expected content '{}'",
-            response_string,
-            expected_error
-        );
+
+        assert_contains!(response_string, "Sugar we are going down");
```
Comment (alamb) [inline src/influxdb_ioxd/rpc/storage/service.rs:1643]: this was a drive by fix -- this file was using several inconsistent patterns to assert that one string was contained in another so I standardized them

File: src/influxdb_ioxd/rpc/storage/service.rs [hunk 14]
```diff
@@ -1022,14 +1022,16 @@ where
         .await
         .context(DatabaseNotFound { db_name })?;
 
-    let executor = db_store.executor();
```
Comment (alamb) [inline src/influxdb_ioxd/rpc/storage/service.rs:1025]: here is the actual replumbing to use the code in the `InfluxRPCPlanner`

**Review by carols10cents — APPROVED** [review]
Overall seems fine, tiny little suggestions to take or leave!

File: query/src/frontend/influxrpc.rs (hunk 4 above)
Comment (NGA-TRAN) [reply query/src/frontend/influxrpc.rs:666]: I think this function will be invoked a lot more in the future, too

File: query/src/frontend/influxrpc.rs [hunk 15]
```diff
@@ -521,23 +540,118 @@ impl InfluxRPCPlanner {
...
+        let mut ss_plans = Vec::with_capacity(table_chunks.len());
+        for (table_name, chunks) in table_chunks {
+            let prefix_columns: Option<&[&str]> = None;
+
+            let ss_plan = self
+                .read_filter_plan(table_name, prefix_columns, &predicate, chunks)
+                .await?;
+            // If we have to do real work, add it to the list of plans
+            if let Some(ss_plan) = ss_plan {
+                ss_plans.push(ss_plan);
```
Comment (NGA-TRAN) [inline query/src/frontend/influxrpc.rs:556]: This is un related question that I just want to know what a plan includes. Let me know if I understand correctly.

1. ss_plan is SeriesSetPlan of a Table.
2. ss_plan's member, plan,  is a DataFusion's LogicalPlan which is built through many forms of PlanBuilder  (e.g. LogicalPlanBuilder, ProviderBuilder, ChunkTableBuilder,....). And when it is a ProviderBuilder which a datatype we define, we add chunks that has the table's data.
3. Also, this LogicalPlan, or more specific LogicalPlanBuilder, provides methods such as sort, limit, join, filter, ... to do add order, limit, ... operators in the plan.

So in short, building a DataFusion's LogicalPlan not only hooks operators in but also needs  to hook its source of data the plan will run on.
I am

File: query/src/frontend/influxrpc.rs [hunk 16]
```diff
@@ -521,23 +540,118 @@ impl InfluxRPCPlanner {
...
+            }
+        }
+
+        Ok(ss_plans.into())
```
Comment (NGA-TRAN) [inline query/src/frontend/influxrpc.rs:560]: At our sync tomorrow, can you explain to me what this .into function here for?

File: query/src/frontend/influxrpc.rs (hunk 5 above)
Comment (NGA-TRAN) [reply query/src/frontend/influxrpc.rs:901]: More bug found and fixed

File: query/src/frontend/influxrpc.rs [hunk 17]
```diff
@@ -763,13 +914,215 @@ impl InfluxRPCPlanner {
...
+    ///
+    /// Equivalent to this SQL query for 'aggregates': sum, count, mean
+    /// SELECT
+    ///   tag1...tagN
+    ///   agg_function(_val1) as _value1
+    ///   ...
+    ///   agg_function(_valN) as _valueN
+    ///   agg_function(time) as time
+    /// GROUP BY
+    ///   group_key1, group_key2, remaining tags,
```
Comment (NGA-TRAN) [inline query/src/frontend/influxrpc.rs:946]: I am a bit confused here. "group_key1, group_key2, remaining tags" = "tag1....tagN"?

Comment (NGA-TRAN) [reply query/src/frontend/influxrpc.rs:946]: I think I got it. They are the same but in different order.

**Review by NGA-TRAN — APPROVED** [review]
Looking great, Andrew. I learned how to build a query plan from this PR

File: query/src/frontend/influxrpc.rs (hunk 15 above)
Comment (alamb) [reply query/src/frontend/influxrpc.rs:556]: > 1. ss_plan is SeriesSetPlan of a Table.

Yes, or more precisely a SeriesSetPlan that generates results for a single table

> 2. ss_plan's member, plan, is a DataFusion's LogicalPlan 

Yes. I would describe the creation as:
[`LogicalPlan`](https://docs.rs/datafusion/3.0.0/datafusion/logical_plan/enum.LogicalPlan.html) is the actual DataFusion query plan (a DAG dataflow graph).

The source node in that plan is a [`TableScan`](https://docs.rs/datafusion/3.0.0/datafusion/logical_plan/enum.LogicalPlan.html#variant.TableScan)  node which scans data and potentially applies filtering. 

A `TableScan` uses a [`TableProvider`](https://docs.rs/datafusion/3.0.0/datafusion/datasource/datasource/trait.TableProvider.html) to provide the schema of the table and produce the actual `RecordBatch`es during execution. The [`TableProvider`] for IOx chunks is created using the `ProviderBuilder`,

> 3. Also, this LogicalPlan, or more specific LogicalPlanBuilder, provides methods such as sort, limit, join, filter, ... to do add order, limit, ... operators in the plan.

Yes that is accurate

> So in short, building a DataFusion's LogicalPlan not only hooks operators in but also needs to hook its source of data the plan will run on.

Yes I think that is accurate

File: query/src/frontend/influxrpc.rs (hunk 16 above)
Comment (alamb) [reply query/src/frontend/influxrpc.rs:560]: Done! And for anyone following along the `into()` call here is creating a `SeriesSetPlan` from a `LogicalPlan`.

File: query/src/frontend/influxrpc.rs (hunk 17 above)
Comment (alamb) [reply query/src/frontend/influxrpc.rs:946]: Yes -- exactly -- the order is different. I will add a comment to this effect.

File: query/src/frontend/influxrpc.rs [hunk 18]
```diff
@@ -763,13 +914,215 @@ impl InfluxRPCPlanner {
...
+    /// ORDER BY
+    ///   group_key1, group_key2, remaining tags
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:948]:
```suggestion
    /// GROUP BY
    ///   group_key1, group_key2, remaining tags,
    /// ORDER BY
    ///   group_key1, group_key2, remaining tags
    ///
    /// Note the columns are the same but in a different order 
    /// for GROUP BY / ORDER BY
```

File: query/src/frontend/influxrpc.rs [hunk 19]
```diff
@@ -973,3 +1326,271 @@ struct TableScanAndFilter {
...
+                }
+                .fail()
+            }
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    let mut new_tag_columns = prefix_map
+        .iter()
+        .map(|&i| tag_columns[i])
+        .collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (alamb) [reply query/src/frontend/influxrpc.rs:1378]: This is a good call -- thanks @carols10cents  -- the Mutable buffer doesn't actually have the column names available so I think that is why I didn't do so originally. 👍
